curvefs/src/metaserver/copyset/copyset_node.cpp

```diff
@@ -516,6 +532,252 @@ std::list<PartitionInfo> CopysetNode::GetPartitionInfoList() {
+            name_.c_str(), peerId_.to_string().c_str(), berror(rc));
+        LOG(ERROR) << status.error_str();
+        return status;
+    }
+
+    return butil::Status::OK();
+}
+
+void CopysetNode::DoAddOrRemovePeer(ConfigChangeType type, const Peer& peer,
+                                    braft::Closure* done) {
+    CHECK(type == ConfigChangeType::ADD_PEER ||
+          type == ConfigChangeType::REMOVE_PEER);
+
+    butil::Status status;
+    auto doneGuard = absl::MakeCleanup([&status, &done]() {
+        if (done) {
+            done->status() = status;
+            done->Run();
+        }
+    });
+
+    braft::PeerId target;
+    int rc = target.parse(peer.address());
+    if (rc != 0) {
+        status = butil::Status(EINVAL, "Peer %s is not valid, copyset: %s",
+                               peer.address().c_str(), name_.c_str());
+        LOG(WARNING) << status.error_str();
+        return;
+    }
+
+    std::vector<braft::PeerId> currentPeers;
+    {
+        std::lock_guard<Mutex> lk(confMtx_);
+        conf_.list_peers(&currentPeers);
+    }
+
+    bool hasTargetPeer = false;
+    for (auto& peer : currentPeers) {
+        if (peer == target) {
+            hasTargetPeer = true;
+            break;
+        }
+    }
+
+    if (type == ConfigChangeType::ADD_PEER && hasTargetPeer) {
+        status.set_error(EEXIST,
+                         "Peer %s is already a member of current copyset: %s",
+                         target.to_string().c_str(), name_.c_str());
+        LOG(WARNING) << status.error_str();
+        return;
+    } else if (type == ConfigChangeType::REMOVE_PEER && !hasTargetPeer) {
+        status.set_error(ENOENT,
+                         "Peer %s is not a member of current copyset: %s",
+                         target.to_string().c_str(), name_.c_str());
+        LOG(WARNING) << status.error_str();
+        return;
+    }
+
+    confChangeMtx_.lock();
+    status = ReadyDoConfChange();
+    if (!status.ok()) {
+        confChangeMtx_.unlock();
+        LOG(WARNING) << status.error_str();
+        return;
+    }
+
+    std::move(doneGuard).Cancel();
+    ongoingConfChange_ = OngoingConfChange(type, peer);
+    OnConfChangeDone* confChangeDone =
+        new OnConfChangeDone(this, done, ongoingConfChange_);
+    confChangeMtx_.unlock();
+
+    return type == ConfigChangeType::ADD_PEER
+               ? raftNode_->add_peer(target, confChangeDone)
+               : raftNode_->remove_peer(target, confChangeDone);
+}
+
+void CopysetNode::AddPeer(const Peer& peer, braft::Closure* done) {
+    return DoAddOrRemovePeer(ConfigChangeType::ADD_PEER, peer, done);
+}
+
+void CopysetNode::RemovePeer(const Peer& peer, braft::Closure* done) {
+    return DoAddOrRemovePeer(ConfigChangeType::REMOVE_PEER, peer, done);
+}
+
+void CopysetNode::ChangePeer(const std::vector<Peer>& newPeers,
+                             braft::Closure* done) {
+    butil::Status status;
+    auto doneGuard = absl::MakeCleanup([&status, &done]() {
```

Comment:
why not use done->release()?

Reply:
I have to set the status of `done` if something wrong.